optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -60,21 +64,31 @@ def gaudi_mpt_attention_forward(
     if past_key_value is not None:
         if len(past_key_value) != 0:
             if token_idx is not None:
+                # copy kv to kv_cache
                 past_key_value[0].index_copy_(2, token_idx - 1, key_states)
                 past_key_value[1].index_copy_(2, token_idx - 1, value_states)
-                key_states = past_key_value[0]
-                value_states = past_key_value[1]
+
+                if cache_idx is not None:
+                    key_states = past_key_value[0][:, :, :cache_idx, :]
+                    value_states = past_key_value[1][:, :, :cache_idx, :]
+                else:
+                    key_states = past_key_value[0]
+                    value_states = past_key_value[1]
             else:
                 key_states = torch.cat([past_key_value[0], key_states], dim=2)
                 value_states = torch.cat([past_key_value[1], value_states], dim=2)
                 past_key_value = [key_states, value_states]
     else:
-        past_key_value = [
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-        ]
-        past_key_value[0][:] = key_states[:]
-        past_key_value[1][:] = value_states[:]
+        if cache_idx is not None:
+            cache_shape = (batch_size, self.n_heads, kv_cache_len, self.head_dim)
+            past_key_value = [
+                torch.empty(cache_shape, dtype=key_states.dtype, device=key_states.device),
+                torch.empty(cache_shape, dtype=key_states.dtype, device=key_states.device),
+            ]
+            past_key_value[0][:, :, : key_states.shape[2], :] = key_states
```

Comment:
is key_states.shape[2] and kv_cache_len different here? aren't they both (input token+output token) during the first token generation?

Reply:
When bucket_internal is enabled(i.e. cache_idx is not None), due to the slice operation [here](https://github.com/Jing1Ling/optimum-habana/blob/86272355a3547504825c70a16c9439a4efab592a/optimum/habana/transformers/models/mpt/modeling_mpt.py#L383), the key_states.shape[2] will be token_idx. Durign first token generation, it is equal to input_token_length.